src/ir/auth_logic/ast_construction.cc

```diff
@@ -78,6 +77,19 @@ static BaseFact ConstructFlatFact(AuthLogicParser::FlatFactContext& context) {
   return BaseFact(ConstructPredicate(*CHECK_NOTNULL(ctx.predicate())));
 }
 
+static BaseFact ConstructRvalue(AuthLogicParser::RvalueContext& context) {
+  if (auto* ctx =
+          dynamic_cast<AuthLogicParser::FlatFactRvalueContext*>(&context)) {
+    return ConstructFlatFact(*CHECK_NOTNULL(ctx->flatFact()));
+  }
+  auto& ctx = *CHECK_NOTNULL(
+      dynamic_cast<AuthLogicParser::BinopRvalueContext*>(&context));
```

Comment:
It looks like you are representing `#binoprvalue`s as predicates where the binary operator is the name of the predicate instead of extending the structure of the AST to have a separate RValue class. I think in principle this is OK, but it will make translating this into souffle more difficult. In the Rust implementation, the compiler pass for 
generating universe relations also makes use of the separate RValue class to figure out that the arguments in the comparison have the `Number` type (https://github.com/google-research/raksha/blob/main/rust/tools/authorization-logic/src/souffle/universe_translation.rs).
I think having an explicit RValue class as part of the AST will make a few other things after the AST is built easier.

For the translation into souffle:
Eventually, when an RVALUE gets emitted as souffle code, it should look like:
 - `pred_name(arg, ..., arg)` if the RVALUE was a predicate, or
 - `arg <binop> arg` if the RVALUE was a binop

So these are different. If you keep your current implementation where a binop is represented by just changing the name of the predicate, you would probably implement the translation with an `if/else` that checks if the name of the predicate is a binary operator. I think this is undoing some of the work that should already have been done by the time the parser produces an AST object. If you have a separate RValue class you would instead implement the translation by visiting the variant in the RVALUE (or, if you make the binop/predicate subclasses of RValue you could use a virtual method instead). I think using an RValue class would result in a more idiomatic translation.

For the compiler pass that populates the universes, you can also distinguish between a binop and a predicate rvalue by checking the predicate name, but I think this is another situation where having an RValue class would make that code simpler.

Reply:
I am just curious why this is resolved. Did you already discuss this option with @bgogul and/or @markww and decide this is not the way to go?

Reply:
Andrew, I thought I provided comments earlier. I didn’t realize I started a review and my comments were in pending state. I just published them. Please take a look and let me know.

Reply:
OK sg. Yes, it was just that the comment was missing as you said :)